README.md
# Lakes380CoreDepthTools

<!-- badges: start -->
[![R-CMD-check](https://github.com/nickmckay/Lakes380CoreDepthTools/workflows/R-CMD-check/badge.svg)](https://github.com/nickmckay/Lakes380CoreDepthTools/actions)
[![R-CMD-check](https://github.com/nickmckay/Lakes380CoreDepthTools/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/nickmckay/Lakes380CoreDepthTools/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

## Installation

This package helps convert core-section depths to depth below lake floor for Lakes 380 lakes. 

To install, you'll need the `remotes` package if you don't already have it

`install.packages("remotes")`

Then you can install the package

`remotes::install_github("nickmckay/Lakes380CoreDepthTools")`

and then load it with

`library(Lakes380CoreDepthTools)`


## How to calculate dblf from depths relative to core liners

Currently, the functionality is pretty basic. To convert core-section depth to depth below lake floor, use the function `coreSection_to_dblf()` with the corename and depth below the top of the core liner. You can run it for one or more depths. Here is an example that will calculate depth below lake floor for 20 to 30 cm below the core liner in core L380_DUNCA3_LC4U_2:


`dblf <- coreSection_to_dblf("L380_DUNCA3_LC4U_2",20:30)`

This will return a dataframe with the depths and some metadata.

All of the depths in the package, both inputs and outputs should be in cm. It should give you an error if you ask for a bad depth range. 


### How to calculate dblf from coreliner depth for multiple core sections.

You may want to convert multiple core sections at once. You can do so using a vector or a data.frame that lists the core section names and corresponding depths. For example:

`multi_coreSection_to_dblf(c("L380_DUNCA3_LC4U_1","L380_DUNCA3_LC4U_2"),c(20,10))`

or in the case where you have the information stored in a data.frame

`multi_coreSection_to_dblf(corec("L380_DUNCA3_LC4U_1","L380_DUNCA3_LC4U_2"),c(20,10))`


### Load in a spreadsheet and add in dblf

The function `dblf_from_file()` will let you select a spreadsheet (csv, xls, xlsx) to load in, select the key columns, and write out a csv file with calculated depths. **This works for both core liner depth, and HSI depth**. Just specify conv.type = "coreliner" or conv.type = "hsi".

If you want to convert multiple cores, try you can add a column with the core name to the spreadsheet, and then select that column in the process. That can look as simple as this:

`dblf_from_file(conv.type = "coreliner")`

Alternatively, if it's just one core, this might be easier (this time for HSI):

`dblf_from_file(corename = "L380_FORSY_LC1U_2",conv.type = "HSI")`



## How to calculate dblf from depths output from hyperspectral imaging (HSI)

HSI depths are typically reported relative to the top of the ROI, not the core liner, so a different function is used to convert those depths, but the pattern is similar: 

`dblf <- hsi_to_dblf("L380_DUNCA3_LC4U_2",20:30)`

This will return a dataframe with the depths and some metadata.

**You'll notice that if you enter a top section, the output depths will be identical to the input depths.** This is by design, since the HSI data typically serve as the depth standard for each core. 


### Similarly to core liner depths, there is an option to convert multiple core sections at once: 

`multi_hsi_to_dblf(c("L380_DUNCA3_LC4U_1","L380_DUNCA3_LC4U_2"),c(20,10))`

or in the case where you have the information stored in a data.frame

`multi_hsi_to_dblf(corec("L380_DUNCA3_LC4U_1","L380_DUNCA3_LC4U_2"),c(20,10))`


## Calculate a core section depth if you know dblf

If you want to do the opposite, use `dblf_to_coreSection()`. Here you input the core name (not the section name) which will be the same, but without the final underscore and section number. For example, we could go backwards in this core by doing:

`cs <- dblf_to_coreSection(corename = df$corename, cm = df$depth)`



## Find the right core name
If you're having trouble getting the right core name, try the helper functions `findCoreSectionName()` where you can input a guess and it will give you suggestions, or `listCoreSectionNames()` to see all the known names. 

There is basic help documentation for these functions, see [reference](https://nickmckay.github.io/Lakes380CoreDepthTools/reference/index.html). 




NEWS.md
# Lakes380CoreDepthTools 0.18.0

* Updated the core depth data file, now current as of 05 Sep 2023

# Lakes380CoreDepthTools 0.17.0

* Reverted the change to `hsi_to_dblf()` in 0.16.0 that was an incorrect/imperfect fix.
* Extended the fix 0.16.0 to depths in `coreSection_to_dblf()` for composite sections with only one core.

# Lakes380CoreDepthTools 0.16.0

* Changed `hsi_to_dblf()` to use maximum (deeper) of compositeRoi top and HSI roi top to avoid situation where composite depth starts above top of HSI depth. 
* changed `coreSection_to_dblf()` to allow multiple composite sections from same core

# Lakes380CoreDepthTools 0.15.3

* Added core section depth and core section to the output of coreSection_to_dblf()
* Added compositing recognition to hsi_to_dblf()

# Lakes380CoreDepthTools 0.15.2

* Modified coreSection_to_dblf to ignore depths outside ROI range for composite cores, and give an warning, rather than stopping with an error. 

# Lakes380CoreDepthTools 0.15.1

* Accommodate blanks in compaction spreadsheet

# Lakes380CoreDepthTools 0.15.0

* Updated the core depth data file, now current as of 22 Apr 2023

# Lakes380CoreDepthTools 0.14.0

* Added the capacity to accommodate more complex or long core composites, and resolve them
* Updated the core depth data file with Adelaine's updates on 22 Mar 2023

# Lakes380CoreDepthTools 0.13.0

* Updated the core depth data file, now current as of 7 Oct 2022

# Lakes380CoreDepthTools 0.12.0

* Updated the core depth data file, now current as of 10 May 2022

# Lakes380CoreDepthTools 0.11.2

* Updated the core depth data file, now current as of 21 April 2022

# Lakes380CoreDepthTools 0.11.1

* Updated the core depth data file, now current as of 29 March 2022

# Lakes380CoreDepthTools 0.11.0

* Updated the core depth data file, now current as of 16 February 2022

# Lakes380CoreDepthTools 0.10.0

* Updated the core depth data file, now current as of 10 February 2022

# Lakes380CoreDepthTools 0.9.0

* Updated the core depth data file, now current as of 29 November 2021

# Lakes380CoreDepthTools 0.8.0

* `dblf_from_file()` now works for HSI too. 
* readme updated

# Lakes380CoreDepthTools 0.7.2

* Updated the core depth data file, now current as of 4 November 2021
* HSI depths now allow depths that are now < 1cm below the bottom of the ROI will be assigned the deepest depth within the ROI. If it is more that 1 cm below, an error will still be generated.

# Lakes380CoreDepthTools 0.7.1

* Added functionality to convert HSI depths


# Lakes380CoreDepthTools 0.7.0

* Updated the core depth data file, now current as of 15 October 2021
* Depths that are now < 1cm below the bottom of the ROI will be assigned the deepest depth within the ROI. If it is more that 1 cm below, an error will still be generated

# Lakes380CoreDepthTools 0.6.0

* Updated the core depth data file, now current as of 7 October 2021

# Lakes380CoreDepthTools 0.5.0

* Added function to account for surface compaction. See `adjustForCompaction` for details.
* Added a function to find core section and core liner depth from dblf. `dblf_to_coreSection`. 

# Lakes380CoreDepthTools 0.4.1

* Fixed bug with hysperspectral data that aren't aligned to zero.

# Lakes380CoreDepthTools 0.4.0

* Updated the core depth data file, now current as of 27 September 2021


# Lakes380CoreDepthTools 0.3.0

* Updated the core depth data file, now current as of 17 September 2021
* Added a `NEWS.md` file to track changes to the package.


R/ui.R
#' Calculate mblf from a file
#'
#' @param filename Optionally enter the file name (default = NA, will let you pick)
#' @param outfile Optionally enter the output file path and name (default = NA, will append "-dblf" and save as csv
#' @param corename core name, using the "L380-" prefix. If NA, you can select a column from the spreadsheet
#' @param conv.type which conversion to use. For now the only option is coreliner
#'
#' @importFrom crayon bold
#' @importFrom readxl read_xls read_xlsx
#' @importFrom readr read_csv write_csv
#' @importFrom tools file_ext file_path_sans_ext
#' @importFrom dplyr bind_cols
#'
#' @return an updated file
#' @export
dblf_from_file <- function(filename = NA,
                           outfile = NA,
                           corename = NA,
                           conv.type = "coreliner"){

  if(is.na(filename)){
    filename <- file.choose()
  }

  ext <- tools::file_ext(filename)
  name <- tools::file_path_sans_ext(filename)
  if(is.na(outfile)){
    outfile <- paste0(name,"-dblf.csv")
  }

  if(ext == "csv"){
    read <- readr::read_csv
  }else if(ext == "xls"){
    read <- readxl::read_xls
  }else if(ext == "xlsx"){
    read <- readxl::read_xlsx
  }else{
    stop("Only set up tp load csv, xls or xlsx files so far.")
  }
  #read in the data
  data_in <- read(filename)
  dn <- names(data_in)#get the data names


  if(is.na(corename)){

    cat(crayon::bold("Select the core name variable:\n"))

    for(i in 1:length(dn)){
      cat(paste(i,"-",dn[i],"\n"))
    }
    n = readline(prompt="please type the number for the correct match, or a zero  to enter a corename ")
    idi=as.numeric(n)

    if(idi == 0){
      cn = readline(prompt="Enter the corename, starting with `L380_`")

    }else{

      cn <- data_in[,idi]
      if(is.list(cn)){
        cn <- unlist(cn)
      }
    }
  }else{
    cn <- corename
  }

  if(conv.type == "coreliner"){
    cat(crayon::bold("Select the depth below coreliner (in cm) variable:\n"))
  }else if(tolower(conv.type) == "hsi"){
    cat(crayon::bold("Select the HSI depth  (in cm) variable:\n"))
  }


  for(i in 1:length(dn)){
    cat(paste(i,"-",dn[i],"\n"))
  }
  n = readline(prompt="please type the number for the correct match, or a zero if none match: ")
  idi=as.numeric(n)

  if(idi == 0){
    stop("You said that none matched")
  }

  dp <- data_in[,idi]
  if(is.list(dp)){
    dp <- unlist(dp)
  }

  cat(crayon::bold("This can take a few seconds to read/write\n"))

  if(tolower(conv.type) == "hsi"){
    if(length(unique(cn)) > 1){
      newDepths <- multi_hsi_to_dblf(cn,dp)
    }else{
      newDepths <- hsi_to_dblf(cn,dp)
    }
  }else{
    newDepths <- multi_coreSection_to_dblf(cn,dp)
  }


  data_out <- dplyr::bind_cols(data_in,newDepths)

  readr::write_csv(data_out,path = outfile)

  cat(crayon::green(paste("Data written to",crayon::bold(outfile))))

  return(paste("Data written to",outfile))

}


R/compact.R
#' Adjust for compaction based on parameters
#'
#' @param origDepth unadjusted depth in cm. Can be negative so long as it's > -maxCompact
#' @param maxCompact Maximum amount of compaction in cm (i.e., the compaction observed at the surface)
#' @param compactOver Depth over which to compensate in the compaction
#' @param eEnd Where to end the exp function scale. Here, a parameter that controls the linearity of the the compaction. Lower values are more linear. Must be greater than 0.  (Default =3)
#' @importFrom glue glue
#' @return adjusted depth
#' @export
#'
#' @examples
#' depths <- seq(0,6,by = .1)
#' ad <- purrr::map_dbl(depths, adjustForCompaction, maxCompact = 0.5, compactOver = 5)
#'

adjustForCompaction <- function(origDepth,maxCompact,compactOver,eEnd = 3){

  #check inputs
  if(origDepth < -maxCompact){
    stop(glue::glue("The supplied depth ({origDepth}) is more negative than the amount of compaction ({maxCompact)} can accommodate"))
  }

  if(eEnd <= 0){stop("eEnd must be greater than 0")}

  if(compactOver <= maxCompact){
    stop(glue::glue("The distance over which to accommodate the compaction ({compactOver}) must be larger than the amount of compaction ({maxCompact})"))
  }

  #factor to scale to eEnd
  cf <- eEnd / (compactOver+maxCompact)


  if(origDepth < compactOver){
    compDepth <- origDepth+maxCompact*exp(-(maxCompact+origDepth)*cf)
  }else{
    compDepth <- origDepth
  }

return(compDepth)

}


data-raw/DATASET.R
library(dplyr)

getRoiTop <- function(secTop,foamTop,hsTop,hsRoiTop){
  if(is.na(foamTop)){foamTop <- 0}
  if(is.na(hsRoiTop)){
    roiTop <- secTop + foamTop
    topSource <- "non-HSI"
  }else{
    if(is.na(hsTop)){
      roiTop <- hsRoiTop - secTop
      topSource <- "HSI"

    }else{
      roiTop <- hsRoiTop - hsTop
      topSource <- "HSI"

    }
  }

  return(bind_cols(roiTop = roiTop,topSource = topSource))
}

getRoiBot <- function(secTop,secLenField,secLenLab,foamBot,hsTop,hsRoiBot){
  if(is.na(foamBot)){foamBot <- 0}
  if(is.na(hsRoiBot)){
    if(is.na(secLenLab)){
    roiBot <- secTop + secLenField - foamBot
    botSource <- "non-HSI"
    }else{
      roiBot <- secTop + secLenLab - foamBot
      botSource <- "non-HSI"
    }
  }else{
    if(is.na(hsTop)){
      roiBot <- hsRoiBot - secTop
      botSource <- "HSI"

    }else{
      roiBot <- hsRoiBot - hsTop
      botSource <- "HSI"

    }
  }

  return(bind_cols(roiBot = roiBot,botSource = botSource))
}

#also extract which (HS or not) was used.
m_to_cm <- function(m){
  return(as.numeric(m) * 100)
}

coreData <- readxl::read_xlsx(
  system.file("extdata", "Lakes380Cores.xlsx", package = "Lakes380CoreDepthTools"),
  col_types = "text",
) %>%
  mutate(across(contains("Section") & -ends_with("Section Name"),m_to_cm)) %>%
  rename(compact = `Compaction off set to correct [cm]`,
         compactOver = `Depth over which to correction compaction [cm]`)

roiTop <- select(coreData,
                 secTop = `Section Top`,
                 foamTop = `Foam Top`,
                 hsTop = `HS Core Liner Top`,
                 hsRoiTop = `HS ROI Top`) %>%
  mutate(across(.fns = as.numeric)) %>%
  purrr::pmap_dfr(.f = getRoiTop)

roiBot <- select(coreData,
                 secTop = `Section Top`,
                 secLenField = `Section Length (Field)`,
                 secLenLab = `Remeasured Section Length (Lab)`,
                 foamBot = `Foam Bottom`,
                 hsTop = `HS Core Liner Top`,
                 hsRoiBot = `HS ROI Bottom`) %>%
  mutate(across(.fns = as.numeric)) %>%
  purrr::pmap_dfr(.f = getRoiBot)


key <- select(coreData,ends_with("Section Name"),starts_with("compact"),) %>%
  bind_cols(roiTop,roiBot) %>%
  mutate(sectionLength = roiBot - roiTop,
         coreName = stringr::str_remove(`Section Name`,"[_\\d{1,}]+$"),
         sectionNumber = as.numeric(stringr::str_extract(`Section Name`,pattern = "[\\d{1,}]+$"))) %>%
  arrange(sectionNumber)


secBotDblf <- secTopDblf <- matrix(NA, nrow = nrow(key))

for(i in 1:nrow(key)){
  if(key$sectionNumber[i] == 1){
    secTopDblf[i] <- 0
  }else{
    prevSecNum <- key$sectionNumber[i]-1
    prevI <- which(key$sectionNumber == prevSecNum & key$coreName == key$coreName[i])
    if(length(prevI) == 0){
      secTopDblf[i] <- NA
    }else{
    secTopDblf[i] <- secBotDblf[prevI]
    }
  }
  secBotDblf[i] <- key$sectionLength[i]+secTopDblf[i]
}

finalKey <- key %>%
  bind_cols(data.frame(secTopDblf = secTopDblf, secBotDblf = secBotDblf)) %>%
  arrange(coreName, sectionNumber)


usethis::use_data(finalKey, overwrite = TRUE,internal = TRUE)


R/coreSectionDepth.R
#' Get core section depth from dblf
#'
#' @param core Name of the core (not the core section!) (e.g. "L380_OTOTO_LC4U")
#' @param dblf depth below the lake floor in cm
#' @importFrom purrr pmap_lgl
#' @importFrom dplyr between filter
#' @importFrom glue glue
#' @return a data.frame with core section and depth
#' @export
#'
#' @examples
#' dblf_to_coreSection("L380_OTOTO_LC4U",110)
#' dblf_to_coreSection("L380_OTOTO_LC1U",10)
dblf_to_coreSection <- function(core,dblf){
  #build a table for all cores

  if(all(is.na(dblf))){
    out <- data.frame(dblf,core,NA)
    out <- setNames(out,c("dblf (cm)","Core Section","Section Depth (cm)"))
    return(out)
  }

  theseCores <- dplyr::filter(finalKey,coreName == core)
  if(nrow(theseCores) < 1 | nrow(theseCores) > 6){
    stop(glue::glue("check your core name, we found {nrow(theseCores)} core sections with a core name of {core}. Did you enter a core and not a section name?"))
  }

  whichCore <- which(
    purrr::pmap_lgl(theseCores,
                    function(secTopDblf,secBotDblf,...){
                      dplyr::between(dblf,left = secTopDblf, right = secBotDblf)
                    }
    )
  )

  if(length(whichCore)<1){
    stop("Dblf is outside the possible range for these cores.")
  }
  if(length(whichCore)>1){
    warning("Multiple cores found with this dblf. This probably means the dblf is on a boundary. Taking the upper core")
    whichCore <- min(whichCore)
  }

  thisCore <- theseCores[whichCore,]
  sds <- seq(thisCore$roiTop,thisCore$roiBot,length.out = 100)
  sdblf <- coreSection_to_dblf(thisCore$`Section Name`,sds)

  secDepth <- stats::approx(sdblf$dblf,sds,dblf)$y

  out <- data.frame(dblf,thisCore$`Section Name`,secDepth)
  out <- setNames(out,c("dblf (cm)","Core Section","Section Depth (cm)"))

  return(out)

}






#' Calculate depth below lake floor
#'
#' @param corename Name of the lakes 380 core section
#' @param cm section depth, from the top of the core liner, in cm
#' @param extraAllowedBottom how many cm below the bottom of the ROI are allowed (converted to bottom of ROI) (default = 1)
#' @param extraAllowedTop how many cm above the top of the ROI are allowed (converted to top of ROI) (default = 0.25)
#' @importFrom glue glue
#' @importFrom dplyr filter
#'
#' @return depth below lake floor in cm
#' @export
#'
#' @examples
#'
#' coreSection_to_dblf("L380_DUNCA3_LC4U_1",50:60)
#'
coreSection_to_dblf <- function(corename,cm,extraAllowedBottom = 1,extraAllowedTop = 0.25){

  if(all(is.na(cm))){
    return(NA)
  }

  if(!is.character(corename)){
    stop("corename should be of class character")
  }

  corename <- unique(corename)
  if(length(corename) > 1){
    stop("more than one corename was entered. To calculate depths from multiple sections, use multi_coreSection_to_dblf")
  }

  whichSection <- rep(1,times = length(cm))

  #check to see if it's part of a master composite
  if(tolower(corename) %in% tolower(finalKey$`Original Section Name`)){
    isComposite <- TRUE
    #determine master corename by name and depth
    section <- dplyr::filter(finalKey,tolower(corename) == tolower(`Original Section Name`))
    whichSection <- c()
    for(r in 1:nrow(section)){
      ts <- which(dplyr::between(cm,section$roiTop[r],section$roiBot[r]))
      if(length(ts) > 0){
        whichSection[ts] <- r
      }
    }

    if(all(is.na(whichSection))){
      stop(glue::glue("Couldn't find any master core sections that match this name and depth range, perhaps you entered the wrong depth(s)"))
    }


  }else{
    #find the relevant core section row
    isComposite <- FALSE

    section <- dplyr::filter(finalKey,tolower(corename) == tolower(`Section Name`))


    if(nrow(section) == 0){
      stop(glue::glue("Couldn't find a core section named {corename}. Search for corenames with `findCoreSectionName`, or for a list of known core sections run `listCoreSectionNames()`"))
    }

    if(nrow(section) > 1){
      stop("Multiple core section matches. This shouldn't happen")
    }
  }

  #find which sections to use:
  uws <- as.numeric(na.omit(unique(whichSection)))
  alldblf <- matrix(NA,nrow = length(cm))
  for(tuws in uws){
    # Get key metadata
    secTopDblf <- section$secTopDblf[tuws]
    secRoiTop <- section$roiTop[tuws]
    secRoiBot <- section$roiBot[tuws]

    #check for compaction adjustment
    compact <- FALSE
    if(!is.na(section$compact[tuws]) & !is.na(section$compactOver[tuws])){
      if(is.numeric(section$compact[tuws]) & is.numeric(section$compactOver[tuws])){
        compact <- TRUE
      }
    }

    tcm <- cm[which(whichSection == tuws)]

    #adjust to top depth if within extraAllowedTop cm.
    if(any(between(tcm - secRoiTop,-extraAllowedTop,0))){
      tc <- which(between(tcm - secRoiTop,-extraAllowedTop,0))
      tcm[tc] <- secRoiTop
    }

    #adjust to bottom depth if within extraAllowedBottom cm.
    if(any(between(secRoiBot - tcm,-extraAllowedBottom,0))){
      tc <- which(between(secRoiBot - tcm,-extraAllowedBottom,0))
      tcm[tc] <- secRoiBot
    }

    if(compact){#check for bottom depth
      if(is.numeric(secRoiBot)){
        if(any(secRoiBot < tcm)){
          badDepth <- tcm[which(secRoiBot < tcm)]
          if(isComposite){
            warning(glue::glue("At least one requested depth ({badDepth[1]} cm) is below the ROI range ({secRoiTop} to {secRoiBot} cm) for core {corename}. This is a composite record, so we'll just ignore those depths for now."))
            tcm <- tcm[-which(secRoiBot < tcm)]
            if(length(tcm) == 0){
              stop("After removing depths outside the range, there were no depths left.")
            }

          }else{
            stop(glue::glue("At least one requested depth ({badDepth[1]} cm) is below the ROI range ({secRoiTop} to {secRoiBot} cm) for core {corename}"))
          }
        }
      }
    }else{#check for both
      if(is.numeric(secRoiBot)){
        if(any(secRoiBot < tcm | secRoiTop > tcm)){
          badDepth <- tcm[which(secRoiBot < tcm | secRoiTop > tcm)]
          if(isComposite){
            warning(glue::glue("At least one requested depth ({badDepth[1]} cm) is outside the ROI range ({secRoiTop} to {secRoiBot} cm) for core {corename}. This is a composite record, so we'll just ignore those depths for now."))
            inRange <- which(secRoiBot < tcm)
            tcm <- tcm[-inRange]

            if(length(tcm) == 0){
              stop("After removing depths outside the range, there were no depths left.")
            }

          }else{
            stop(glue::glue("At least one requested depth ({badDepth[1]} cm) is outside the ROI range ({secRoiTop} to {secRoiBot} cm) for core {corename}"))
          }
        }
      }
    }


    if(!is.numeric(secTopDblf) | !is.numeric(secRoiTop)){
      stop("missing metadata for core {corename}, cannot calculate depth below lake floor")
    }

    #then calculate
    dblf <- tcm - secRoiTop + secTopDblf #add in modifier for difference between HSI ROI and chron ROI?

    if(compact){
      compacted <- c()
      for(icm in 1:length(tcm)){
        if(any(is.na(suppressWarnings(as.numeric(section$compactOver[tuws]))))){
          compactThis <- FALSE
        }else{
          if(dblf[icm] > as.numeric(section$compactOver[tuws])){
            compactThis <- FALSE
          }else{
            compactThis <- TRUE
          }
        }

        #compact if necessary
        if(compactThis){
          dblf[icm] <- adjustForCompaction(dblf[icm],
                                           maxCompact = as.numeric(section$compact[tuws]),
                                           compactOver = as.numeric(section$compactOver[tuws])
          )
        }

        compacted[icm] <- compactThis
      }
    }else{
      compacted <- FALSE
    }
    alldblf[which(whichSection == tuws)] <- dblf

  }
  return(data.frame(corename = corename,
                    coreSectionDepth = cm,
                    dblf = alldblf,
                    topSource = section$topSource[1],
                    botSource = section$botSource[1],
                    coreName = corename,
                    compactionAdjusted = compacted,
                    compositeSequence = isComposite))

}


#' Convert depths for multiple core sections
#'
#' @param corename a vector of corenames
#' @param cm a vector of corresponding depths that matches the length of corenames
#'
#' @importFrom purrr map2_dfr
#' @return a tibble
#' @export
#'
#' @examples
#'
#' multi_coreSection_to_dblf(c("L380_DUNCA3_LC4U_1","L380_DUNCA3_LC4U_2"),c(20,10))
#'
multi_coreSection_to_dblf <- function(corename,cm){


  if(is.list(corename)){
    corename <- unlist(corename)
  }


  if(is.list(cm)){
    cm <- unlist(cm)
  }

  return(purrr::map2_dfr(corename,cm,.f = coreSection_to_dblf))
}





#' Convert HSI depths for multiple core sections
#'
#' @param corename a vector of corenames
#' @param cm a vector of corresponding depths that matches the length of corenames
#'
#' @importFrom purrr map2_dfr
#' @return a tibble
#' @export
#'
#' @examples
#'
#' multi_hsi_to_dblf(c("L380_DUNCA3_LC4U_1","L380_DUNCA3_LC4U_2"),c(20,10))
#'
multi_hsi_to_dblf <- function(corename,cm){


  if(is.list(corename)){
    corename <- unlist(corename)
  }


  if(is.list(cm)){
    cm <- unlist(cm)
  }

  return(purrr::map2_dfr(corename,cm,.f = hsi_to_dblf))
}




#' List known core names
#'
#' @return a list of known core names
#' @export
#'
#' @examples
#' allNames <- listCoreSectionNames()
listCoreSectionNames <- function(){
  return(finalKey$`Section Name`)
}



#' Find a core section name
#'
#' @param corename_guess A string that is a guess at a corename
#'
#' @return Potential matches of core names given a guess
#' @export
#'
#' @examples
#' findCoreSectionName("ngano")
findCoreSectionName <- function(corename_guess){
  index <- agrep(pattern = corename_guess,x = finalKey$`Section Name`,ignore.case = TRUE)
  if(length(index) == 0){
    stop("No matches identified. Try the 5 or 6 character lake code name, e.g. `DUNCA`.")
  }
  if(length(index) > 50){
    warning("You found more than 50 matches, maybe narrow down your search?")
  }
  return(finalKey$`Section Name`[index])
}


#' Convert depths from hyperspectral imaging
#'
#' @param corename a vector of corenames
#' @param cm a vector of corresponding depths that matches the length of corenames
#' @param extraAllowedBottom how many cm below the bottom of the ROI are allowed (converted to bottom of ROI) (default = 1)
#' @return depth below lake floor in cm
#' @export
#'
#' @examples
#'
#' hsi_to_dblf("L380_DUNCA3_LC4U_2",0:30)
#'
hsi_to_dblf <- function(corename,cm,extraAllowedBottom = 1){

  if(all(is.na(cm))){
    return(NA)
  }

  if(!is.character(corename)){
    stop("corename should be of class character")
  }

  corename <- unique(corename)
  if(length(corename) > 1){
    stop("more than one corename was entered. To calculate depths from multiple sections, use multi_hsi_to_dblf")
  }

  #find the relevant core section row
  #check to see if it's part of a master composite
  if(tolower(corename) %in% tolower(finalKey$`Original Section Name`)){
    isComposite <- TRUE
    #determine master corename by name and depth
    sectionComposite <- dplyr::filter(finalKey,tolower(corename) == tolower(`Original Section Name`))
    section <- dplyr::filter(finalKey,tolower(corename) == tolower(`Section Name`))

    # Get key metadata
    secTopDblf <- section$secTopDblf[1]
    secRoiTop <- section$roiTop[1]
    secRoiBot <- section$roiBot[1]
    # if(nrow(section) > 1){
    #   goodRow <- c()
    #   for(r in 1:nrow(section)){
    #     goodRow[r] <- all(between(cm,section$roiTop[r],section$roiBot[r]))
    #   }
    #
    #   if(sum(goodRow) == 1){#we found it!
    #     section <- section[goodRow,]
    #   }else if(sum(goodRow) == 0){
    #     stop(glue::glue("Couldn't find any master core sections that match this name and depth range, perhaps you entered the wrong depth(s)"))
    #   }else{
    #     stop(glue::glue("Found multiple master core sections that match this name and depth range, this seems like a problem with the depth table."))
    #   }
    #
    # }

  }else{
    #find the relevant core section row
    isComposite <- FALSE

    section <- dplyr::filter(finalKey,tolower(corename) == tolower(`Section Name`))

    # Get key metadata
    secTopDblf <- section$secTopDblf[1]
    secRoiTop <- section$roiTop[1]
    secRoiBot <- section$roiBot[1]
  }

  if(nrow(section) == 0){
    stop(glue::glue("Couldn't find a core section named {corename}. Search for corenames with `findCoreSectionName()`, or for a list of known core sections run `listCoreSectionNames()`"))
  }

  if(nrow(section) > 1){
    stop("Multiple core section matches. This shouldn't happen")
  }




  #convert to core liner
  clDepth <- cm + secRoiTop

  # #adjust to bottom depth if within 1 cm.
  # if(any(between(secRoiBot - clDepth,-extraAllowedBottom,0))){
  #   tc <- which(between(secRoiBot - clDepth,-extraAllowedBottom,0))
  #   clDepth[tc] <- secRoiBot
  # }


  #try to get dblf
  dblf <- coreSection_to_dblf(corename,clDepth,extraAllowedBottom = extraAllowedBottom)

  return(dblf)

  #
  #
  # #convert back to HS depth
  # cm <- clDepth - secRoiTop
  #
  # if(is.numeric(secRoiBot) & is.numeric(secRoiTop)){
  #   if(any((secRoiBot - secRoiTop) < cm)){
  #     badDepth <- cm[which((secRoiBot - secRoiTop) < cm)]
  #     stop(glue::glue("At least one requested depth ({badDepth[1]} cm on HSI scale, {badDepth[1] + secRoiTop} relative to the coreliner) is outside the ROI range ({secRoiTop} to {secRoiBot} cm) for core {corename}"))
  #   }
  # }
  #
  #
  #
  # if(!is.numeric(secTopDblf)){
  #   stop("missing metadata for core {corename}, cannot calculate depth below lake floor")
  # }
  #
  # #then calculate
  # dblf <- cm + secTopDblf
  #
  #   return(data.frame(dblf = dblf,
  #                     topSource = section$topSource[1],
  #                     botSource = section$botSource[1],
  #                     coreName = corename))

}

